packages/server/src/manifest-builder.ts

```diff
@@ -77,59 +76,66 @@ function* processManifest(options: ManifestBuilderOptions): Operation {
   let manifest = yield import(distPath);
 
   manifest = manifest.default || manifest;
-
   manifest.fileName = fileName;
 
-
   let slice = options.atom.slice('manifest');
+  
   slice.set(manifest as Manifest);
 
   return distPath;
 }
 
-function logBuildError(error: BundlerError) {
-  console.error("[manifest builder] build error:", error.message);
-  if (error.frame) {
-    console.error("[manifest builder] build error frame:\n", error.frame);
-  }
-}
-
-function* waitForSuccessfulBuild(bundlerEvents: ChainableSubscription<BundlerMessage, undefined>, delegate: Mailbox): Operation {
-  let message: BundlerMessage = yield bundlerEvents.expect();
-
-  if (message.type === "error") {
-    logBuildError(message.error);
-    delegate.send({ event: "error" });
-    yield waitForSuccessfulBuild(bundlerEvents, delegate);
-  }
-}
-
 export function* createManifestBuilder(options: ManifestBuilderOptions): Operation {
+  let bundlerSlice = options.atom.slice('bundler');
+
+  bundlerSlice.set({ kind: 'UNBUNDLED' });
+  
   let bundler: Bundler = yield Bundler.create(
     [{
       entry: options.srcPath,
       globalName: bigtestGlobals.manifestProperty,
       outFile: path.join(options.buildDir, "manifest.js")
-    }]
+    }],
   );
-  let bundlerEvents: ChainableSubscription<BundlerMessage, undefined> = yield subscribe(bundler);
-
-  yield waitForSuccessfulBuild(bundlerEvents, options.delegate);
 
-  let distPath: string = yield processManifest(options);
-
-  console.debug("[manifest builder] manifest ready");
-
-  options.delegate.send({ status: "ready", path: distPath });
-
-  yield bundlerEvents.forEach(function*(message) {
-    if (message.type === 'error') {
-      logBuildError(message.error);
-      options.delegate.send({ event: "error" });
-    } else {
-      let distPath = yield processManifest(options);
-      console.info("[manifest builder] manifest updated");
-      options.delegate.send({ event: "update", path: distPath });
+  yield Subscribable.from(bundler).forEach(function* (message) {
+    switch (message.kind) {
+      case 'START':
+        console.debug("[manifest builder] received bundler start");
+        
+        bundlerSlice.update(() => ({ kind: 'BUILDING', warnings: [] }));
+        break;
+      case 'UPDATE':
+        console.debug("[manifest builder] received bundle update");
+        
+        let path: string = yield processManifest(options);
+        
+        bundlerSlice.update((previous) => {
+          // as this is a typescript assertion it does more than just check the condition
+          // it will type narrow the discriminated union based on the kind discriminator
+          assert(previous.kind === 'BUILDING', `invalid transition from ${previous.kind} to 'GREEN'`);
```

Comment:
oh I see, here we use it. Should this be a warning?

Reply:
So this does more than just throw if the condition is not met.  If the `previous.kind === 'BUILDING;`

Then I can do this

`return { ...previous, kind: 'GREEN', path };`

the `path property` only exists when the  `kind` or `type` is green.  the `assert` here narrows the type. 

If `previous.kind !== 'BUILDING;'` then it is cast to `never`.

It should not be its own package because it is just using a typescript feature we are building upon.

Reply:
I put it in `@bigtest/project` because I thought that might be where code that is used by all project goes.  I